lib/introspect.js

```diff
@@ -20,32 +20,68 @@ module.exports = function(obj, callback) {
               ifaceName = iface['@'].name;
               currentIface = proxy[ifaceName] = {};
 
-              currentIface.addListener = currentIface.on = (function(ifName) {
-                  return function(signame, callback) {
-                      // TODO: check if AddMatch had already been called
-                      // http://dbus.freedesktop.org/doc/api/html/group__DBusBus.html#ga4eb6401ba014da3dbe3dc4e2a8e5b3ef
-                      // An example is "type='signal',sender='org.freedesktop.DBus', interface='org.freedesktop.DBus',member='Foo', path='/bar/foo',destination=':452345.34'" ...
-                      //var match = "type='signal',sender='"+obj.service.name+ ...
-                      
-                      //TODO: add path and interface to path
-                      var match = "type='signal',member='" + signame + "'";
-                      bus.addMatch(match, function(err, result) {
-                          if (err) throw new Error(err);
-                          // TODO: share signal mangling code
-                          var signalFullName = bus.mangle(obj.name, ifName, signame); 
-                          //console.log('trying to listen' +  signalFullName, obj);
-                          bus.signals.on(signalFullName, function(messageBody)
-                          {
+              (function() {
+                  var callbacks = [];
+                  var sigHandlers = [];
+                  function getSigHandler(callback) {
+                      var index;
+                      if ((index = callbacks.indexOf(callback)) == -1) {
+                          index = callbacks.push(callback) - 1;
+                          sigHandlers[index] = function(messageBody) {
                                callback.apply(null, messageBody)
-                          });
```

Comment:
Umm. No, it is not the same.

The second parameter in apply is an array, which when function called goes as separate arguments.

For example, `callback.apply(null, [1,2,3])` is equivalent to `callback(1,2,3)`

If we do `sigHandlers[index] = callback`, API will change and it'll go as `callback([1,2,3])`